### ISSUE TITLE
Default-Cookie-Experiation-Time only works 

### ISSUE THREAD

**Issue by GolexPrime** [simplabs/ember-simple-auth#395]
Hi, 
I use the cookie store. I give the user the option to choose if he wants a session until the browser closes, or if he wants a 2 week session (expiration time: 1209600). This works fine if the user closes the browser just one time.

Let me explain:
After he closes the browser the first time, the cookie remains until he visits the application a second time. Because he doesnt have to log in, the expiration time wont be set to 2 weeks. If he closes the browser again the session (and the cookie) will be lost. 

Does the default-value of the expiration-time affect an existing cookie?

**Comment by marcoow** [comment]
How do you implement that? I guess you're using the cookie session store for that?

**Comment by GolexPrime** [comment]
We're using a custom authenticator. In the LoginController I if a certain checkbox is activated. If thats the case I execute:

this.set('session.store.cookieExpirationTime', 1209600);

Thats the only time I set the expiration-time (in the authenticate-action of the LoginController).

**Comment by marcoow** [comment]
You need to set the expiration time back to its default value (or `null` if you want to use session cookies) whe the checkbox is unchecked.

**Comment by GolexPrime** [comment]
Here is the authenticate-Function:

actions: {
		authenticate: function() {

			var _this 		        = this;
			var rememberMe 	= $("#remember-me");

			this._super().then(null, function(message) {
				_this.set('errorMessage', message);
			});

			if(rememberMe.prop("checked")) {
				this.set('session.store.cookieExpirationTime', 1209600);
			} else {
				this.set('session.store.cookieExpirationTime', null);
			}		
		}
	}

My Problem is that when I close the browser (after the login) and open it again the cookie still there. But when I close the Browser a second time, the cookie is gone.

**Comment by marcoow** [comment]
Try setting the expiration time *before* you call `_super`.

**Comment by GolexPrime** [comment]
Just tried it, still the same problem.

**Comment by GolexPrime** [comment]
I think the problem is that the expiration-time of the cookie is set to default (null in my case) when the browser is opened again. But I may be wrong of course.

**Comment by marcoow** [comment]
Can you post screenshots of Developer Tools -> Resources -> Cookies at the different points in time?

**Comment by GolexPrime** [comment]
Ok, 

before I log in:
![bildschirmfoto 2015-01-12 um 10 03 49](https://cloud.githubusercontent.com/assets/4212661/5700740/b29637ee-9a42-11e4-8a9f-debd1ceef42f.png)

after I logged in:
![bildschirmfoto 2015-01-12 um 10 04 38](https://cloud.githubusercontent.com/assets/4212661/5700745/c080c3ba-9a42-11e4-8dee-79179e122d23.png)

after I closed and reopend the browser:
![bildschirmfoto 2015-01-12 um 10 05 25](https://cloud.githubusercontent.com/assets/4212661/5700751/d7455c64-9a42-11e4-8571-0eb18b3ba9c0.png)

**Comment by marcoow** [comment]
Is the checkbox checked or unchecked in that example? Did you configure a default cooke expiration time?

**Comment by GolexPrime** [comment]
In this example the checkbox was checked. No, I didn't change the default-value of the expiration-time. so it should be null.

**Comment by marcoow** [comment]
I think I got it. The store gets rewritten after the session gets restored on application startup but because the default expiration time is `null` it will use that then. You'd have to store the expiration time somewhere else as well and set it accordingly in an initializer that runs before the `'simple-auth'` initializer. Not sure whether or how this could be fixed in the library itself actually.

**Comment by GolexPrime** [comment]
Ok, that was my impression as well. 

My Problem was that i didn't found a way to read the expiration time in the cookie. And I didn't want to write it as additional information in the cookie itself because there it would be easy for the user to change it. I'm sure I will find a solution.

Thanks for the effort.

**Comment by marcoow** [comment]
I don't think it's a problem if the user can change the expiration easily because if the user wouldn't be allowed to access/modify the security-critical data in the cookie that would be a problem anyway, but one you couldn't do anything against.

It would probably help if the cookie store wouldn't rewrite the cookie when the cookie already has the exact same data that would be written.

**Comment by GolexPrime** [comment]
It would be awesome if the cookie-store wouldn't overwrite the date at all (if a cookie already exists)^^

**Comment by marcoow** [comment]
It has to override because data may have changed - it just shouldn't replace with the exact same data.

**Comment by GolexPrime** [comment]
That may be a good idea for the future^^

**Comment by snovity** [comment]
> And I didn't want to write it as additional information in the cookie itself because there it would be easy for the user to change it.

For user changing the actual expiration time of a cookie is just as easy as changing its contents. I think, writing expiration time inside cookie doesn't create any additional security risks.

Also, solution with not touching cookie at all may not work as expected, because it will logout user after two weeks even if he used the app on the last day of those two weeks. If one sets a Rails session expiration time to say one week, it will be one week since the last request, but not since the last login.

**Comment by marcoow** [comment]
Good point @snovity.

**Comment by marcoow** [comment]
@snovity, @GolexPrime: would either of you (or anyone else who reads this of course) have time to look into this? I'm quite busy currently...

**Comment by lolmaus** [comment]
Stumbled over this issue. It's a legit bug because the 8th example simply won't persist more than one closing-opening of the browser.

Here's my workaround:

Install https://www.npmjs.com/package/ember-cli-cookie

Add this to your login controller:

```coffee
  # A property the remember me checkbox is bound to
  rememberMe: true

  rememberMeChanged: Ember.observer 'rememberMe', Ember.on 'init', ->
    rememberMe               = @get 'rememberMe'
    Cookie                   = @get 'cookie'
    cookieName               = 'simple-auth-remember-me-expires-on'
    cookie                   = Cookie.getCookie cookieName
    cookieExpirationDuration = null

    if rememberMe and not cookie
      nowSeconds               = Math.floor Date.now() / 1000
      expiresInDays            = 31
      cookieExpirationDuration = 60 * 60 * 24 * expiresInDays #+ nowSeconds
      cookieExpirationTime     = nowSeconds + cookieExpirationDuration
      Cookie.setCookie cookieName, cookieExpirationTime, expires: expiresInDays, path: '/'

    else
      Cookie.removeCookie cookieName

    @get 'session.store'
      .cookieExpirationTime = cookieExpirationDuration
```

Create an initializer `simple-auth-cookie-expiration`:

```coffee
`import CookieStoreConfig from 'simple-auth-cookie-store/configuration'`

# Takes two parameters: container and app
initialize = (container, app) ->
  console.log 'CookieStoreConfig', CookieStoreConfig
  app.inject('controller', 'cookie', 'cookie:main')

  Cookie = container.lookup 'cookie:main'

  # ember-cli-cookie does not accept options. :(
  # Remove this block when this PR is accepted:  https://github.com/achambers/ember-cli-cookie/pull/7
  Cookie.setCookie = (key, value, options) ->
    new Em.RSVP.Promise (resolve, reject) ->
      try
        Em.$.cookie key, value, options
        Em.run null, resolve
      catch e
        Em.run null, reject, e

  Object.defineProperty CookieStoreConfig, 'cookieExpirationTime', set: ->
  Object.defineProperty CookieStoreConfig, 'cookieExpirationTime', get: ->
    rememberMeExpiresOn = Cookie.getCookie 'simple-auth-remember-me-expires-on'

    result =
      if rememberMeExpiresOn?
        rememberMeExpiresOn = parseInt rememberMeExpiresOn, 10
        nowSeconds = Math.floor Date.now() / 1000
        rememberMeExpiresOn - nowSeconds
      else
        null

    result


SimpleAuthCookieExpirationInitializer =
  name: 'simple-auth-cookie-expiration'
  after: ['cookie']
  initialize: initialize

`export {initialize}`
`export default SimpleAuthCookieExpirationInitializer`
```

IE9+. Use on your own risk.

**Comment by marcoow** [comment]
@lolmaus: cool. Do you think you could prepare a PR that changes the cookie store so that all of this would happen automatically inside the store (and probably the initializer of the cookie store extension library)?

**Comment by lolmaus** [comment]
@marcoow, this workaround is messy, hacky, untested and lacks support for old browsers.

Instead of going this way, you really should have session store maintain cookie expiration dates when it recreates them (why does it destroy original cookies in the first place?!). As it is impossible to read cookie expiration date, you should duplicate expiration dates within cookie bodies.

**Comment by marcoow** [comment]
@lolmaus: what you're describing is actually what I was talking about for the PR - transparently storing the expiration date in the cookie and setting/updating it accordingly...

**Comment by lolmaus** [comment]
@marcoow, i'm not familiar with Simple Auth source code. Why don't you do it yourself?

**Comment by marcoow** [comment]
@lolmaus because this is OSS and I encourage people to participate - also I have other things to do as well and am not working on this full time.

**Comment by lolmaus** [comment]
@marcoow, i shared what i managed to do so far.

**Comment by msalahz** [comment]
@lolmaus, Can you please review my PR #451 regarding this issue ?

**Comment by marcoow** [comment]
fixed by #451.
